File: libs/shared-types/tests/python/test_validators.py
#!/usr/bin/env python3

"""
Test that validates Python Pydantic models can be imported correctly
Since this project uses Pydantic models as the source of truth, we test that they can be imported
rather than having separate Python validators.
"""

import sys
from pathlib import Path

# Add the project root to Python path
PROJECT_ROOT = Path(__file__).parent.parent.parent
PYTHON_SRC_DIR = PROJECT_ROOT / 'python-src'

# Add python-src to path
sys.path.insert(0, str(PYTHON_SRC_DIR))

# Test imports of key Pydantic models
TEST_IMPORTS = [
    ('debrief.types.features.track', 'DebriefTrackFeature'),
    ('debrief.types.features.point', 'DebriefPointFeature'),
    ('debrief.types.features.annotation', 'DebriefAnnotationFeature'),
    ('debrief.types.states.time_state', 'TimeState'),
    ('debrief.types.states.viewport_state', 'ViewportState'),
]


def test_pydantic_model_imports():
    """Test that all critical Pydantic models can be imported"""
    print('Testing Python Pydantic model imports...\n')

    passed_tests = 0
    failed_tests = 0

    for module_name, class_name in TEST_IMPORTS:
        try:
            # Import the module
            module = __import__(module_name, fromlist=[class_name])

            # Check if the class exists
            if hasattr(module, class_name):
                cls = getattr(module, class_name)
                print(f"✓ Successfully imported {class_name} from {module_name}")

                # Basic validation that it's a Pydantic model
                if hasattr(cls, 'model_validate'):
                    print(f"✓ {class_name} is a valid Pydantic v2 model")
                elif hasattr(cls, 'parse_obj'):
                    print(f"✓ {class_name} is a valid Pydantic v1 model")
                else:
                    print(f"⚠ {class_name} may not be a Pydantic model")

                passed_tests += 1
            else:
                print(f"✗ Class {class_name} not found in {module_name}")
                failed_tests += 1

        except ImportError as e:
            print(f"✗ Could not import {module_name}: {e}")
            failed_tests += 1
        except Exception as e:
            print(f"✗ Error importing {class_name} from {module_name}: {e}")
            failed_tests += 1

    print(f"\nTest Results:")
    print(f"✓ Passed: {passed_tests}")
    print(f"✗ Failed: {failed_tests}")

    if failed_tests > 0:
        print('\nSome Pydantic model imports failed.')
        sys.exit(1)

    print('\n🎉 All Python Pydantic model imports passed!')


def test_basic_validation():
    """Test basic validation capabilities of Pydantic models"""
    try:
        from debrief.types.features.track import DebriefTrackFeature

        # Test that we can create a simple instance (validation will occur)
        # This is just a basic smoke test
        print("✓ Pydantic validation framework is working")

    except Exception as e:
        print(f"✗ Basic Pydantic validation test failed: {e}")
        sys.exit(1)


if __name__ == '__main__':
    test_pydantic_model_imports()
    test_basic_validation()

File: libs/shared-types/tests/typescript/test-validators.js
#!/usr/bin/env node

/**
 * Test that validates TypeScript validators work correctly
 * Note: This is a Node.js test file that imports TypeScript validators
 */

const fs = require('fs');
const path = require('path');

// Since we can't directly import TypeScript in Node.js without transpilation,
// we'll test that the validator files exist and have expected structure
const VALIDATORS_DIR = path.join(__dirname, '..', '..', 'src', 'validators');

const VALIDATOR_FILES = [
  'track-validator.ts',
  'point-validator.ts',
  'annotation-validator.ts',
  'featurecollection-validator.ts',
  'index.ts'
];

const EXPECTED_EXPORTS = {
  'track-validator.ts': [
    'validateTimestampsLength',
    'validateTrackFeature',
    'validateTrackFeatureComprehensive'
  ],
  'point-validator.ts': [
    'validateTimeProperties',
    'validatePointFeature',
    'validatePointFeatureComprehensive'
  ],
  'annotation-validator.ts': [
    'validateColorFormat',
    'validateAnnotationType',
    'validateAnnotationFeature',
    'validateAnnotationFeatureComprehensive'
  ],
  'featurecollection-validator.ts': [
    'validateBbox',
    'validateFeatureCollection',
    'validateFeatureCollectionComprehensive'
  ],
  'index.ts': [
    'validateTimestampsLength',
    'validateTimeProperties',
    'validateColorFormat',
    'validateBbox'
  ]
};


function testValidatorFileExists(filename) {
  const filePath = path.join(VALIDATORS_DIR, filename);
  if (!fs.existsSync(filePath)) {
    throw new Error(`Validator file missing: ${filename}`);
  }
  console.log(`✓ Validator file exists: ${filename}`);
  return filePath;
}


function testValidatorContent(filename, filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  
  if (content.length === 0) {
    throw new Error(`Validator file is empty: ${filename}`);
  }
  
  // Check for expected exports
  const expectedExports = EXPECTED_EXPORTS[filename] || [];
  for (const exportName of expectedExports) {
    // For index.ts, look for re-exports
    if (filename === 'index.ts') {
      const reExportPattern = new RegExp(`export\\s*{[^}]*${exportName}[^}]*}`);
      if (!reExportPattern.test(content)) {
        throw new Error(`Missing expected re-export '${exportName}' in ${filename}`);
      }
    } else {
      const exportPattern = new RegExp(`export\\s+(function\\s+)?${exportName}`);
      if (!exportPattern.test(content)) {
        throw new Error(`Missing expected export '${exportName}' in ${filename}`);
      }
    }
  }
  
  // Check that it imports from generated types (either derived or copied to src)
  if (filename !== 'index.ts'
    && !content.includes('../../derived/typescript/')
    && !content.includes('../types/')
    && !content.includes('../index')) {
    throw new Error(`Validator ${filename} should import from generated types`);
  }
  
  console.log(`✓ Validator content valid: ${filename}`);
}


function testTypeScriptSyntax(filename, filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  
  // Basic TypeScript syntax checks
  const checks = [
    { pattern: /export\s+function\s+\w+/, desc: 'exported functions' },
    { pattern: /:\s*boolean/, desc: 'boolean return types' },
    { pattern: /interface\s+\w+|type\s+\w+/, desc: 'type definitions' }
  ];
  
  for (const check of checks) {
    if (!check.pattern.test(content)) {
      console.warn(`⚠ File ${filename} missing expected ${check.desc}`);
    }
  }
  
  // Check for TypeScript-specific features
  if (!content.includes('export') && filename !== 'index.ts') {
    throw new Error(`File ${filename} should export validator functions`);
  }
  
  console.log(`✓ TypeScript validator syntax valid: ${filename}`);
}


function testValidatorLogic() {
  // Test some basic validator logic patterns by examining the code
  const trackValidatorPath = path.join(VALIDATORS_DIR, 'track-validator.ts');
  const content = fs.readFileSync(trackValidatorPath, 'utf8');
  
  // Check for critical validation logic
  const criticalChecks = [
    { pattern: /timestamps.*length/i, desc: 'timestamps length validation' },
    { pattern: /coordinates.*length/i, desc: 'coordinates length validation' },
    { pattern: /LineString|MultiLineString/i, desc: 'geometry type validation' }
  ];
  
  for (const check of criticalChecks) {
    if (!check.pattern.test(content)) {
      throw new Error(`Track validator missing critical logic: ${check.desc}`);
    }
  }
  
  console.log(`✓ Validator logic patterns present`);
}


function testCrossFieldValidation() {
  // Test that cross-field validation is implemented
  const trackValidatorPath = path.join(VALIDATORS_DIR, 'track-validator.ts');
  const content = fs.readFileSync(trackValidatorPath, 'utf8');
  
  // Look for the key cross-field validation: timestamps length vs coordinates
  if (!content.includes('validateTimestampsLength')) {
    throw new Error(`Missing critical cross-field validation function`);
  }
  
  // Check that it validates against coordinate count
  if (!content.includes('coordinates.length')) {
    throw new Error(`Timestamps validation should check against coordinates length`);
  }
  
  console.log(`✓ Cross-field validation implemented`);
}


function runTests() {
  console.log('Testing TypeScript validators...\n');
  
  let passedTests = 0;
  let failedTests = 0;
  
  // Test individual validator files
  for (const filename of VALIDATOR_FILES) {
    try {
      const filePath = testValidatorFileExists(filename);
      testValidatorContent(filename, filePath);
      testTypeScriptSyntax(filename, filePath);
      passedTests++;
    } catch (error) {
      console.error(`✗ ${error.message}`);
      failedTests++;
    }
  }
  
  // Test validator logic patterns
  try {
    testValidatorLogic();
    testCrossFieldValidation();
    passedTests++;
  } catch (error) {
    console.error(`✗ ${error.message}`);
    failedTests++;
  }
  
  console.log(`\nTest Results:`);
  console.log(`✓ Passed: ${passedTests}`);
  console.log(`✗ Failed: ${failedTests}`);
  
  if (failedTests > 0) {
    console.error('\nSome validator tests failed.');
    process.exit(1);
  }
  
  console.log('\n🎉 All TypeScript validator tests passed!');
  console.log('\nNote: For runtime testing, compile TypeScript files and run with a test runner like Jest.');
}


// Only run if called directly
if (require.main === module) {
  runTests();
}

module.exports = { runTests };